File: htdbtvlive.py
#!/usr/bin/python

import subprocess, threading, sys, time, string, os, _pg


device = sys.argv[1]
tuner = sys.argv[2]
station = sys.argv[3]
unique = sys.argv[4]

htdb_db = _pg.connect(host="localhost", user="htdb", passwd="htdb", dbname="htdb")

#query = "update tv_recording set recordstatus = 'recording' where id = "+rec_id+";"
#print query
#hdresult = htdb_db.query(query)

live_cmd = "/usr/bin/ffmpeg -i \"http://"+device+":5004/auto/v"+station+"?transcode=mobile\" -f hls /var/www/html/htdb/htdb-tvlive/hdhomerun_"+station+"_"+unique+".m3u8"

class Command(object):
	def __init__(self, cmd):
		self.cmd = cmd
		self.process = None

	def run(self, timeout):
		print "running thread for ", timeout, " seconds"
		def target():
#			print 'Thread started'
			self.process = subprocess.Popen(self.cmd, shell=True)
			self.process.communicate()
#			print 'Thread finished'

		thread = threading.Thread(target=target)
		thread.start()

		thread.join(timeout)
		if thread.is_alive():
			print 'Terminating process'
			self.process.terminate()
			thread.join()
		print self.process.returncode


print "saving stream of "+station
print live_cmd
os.system(live_cmd)

tuner_id = "tuner_one"
if tuner == "tuner1":
	tuner_id = "tuner_two"

#query = "update hdhomerun_devices set "+tuner_id+" = 'off', tuner_used = tuner_used-1 where ipaddress = '"+device+"';"
#print query
#hdresult = htdb_db.query(query)

htdb_db.close()

File: htdbtvlive.php
<?php
//header("Cache-Control: no-cache, must-revalidate"); //HTTP 1.1
//header("Pragma: no-cache"); //HTTP 1.0
//header("Expires: Sat, 26 Jul 1997 05:00:00 GMT"); // Date in the past


$server = $_SERVER['SERVER_ADDR'];
$server_name = $_SERVER['SERVER_NAME'];
print "<!-- '$server', '$server_name' -->\n";
if ($server_name != $server) {$server = $server_name;}

$cookie_client = $_COOKIE["htdb-client"];

$station = $_GET['station'];

if (strlen($station) == 0) {
	$station = "20.2";
}
if (strlen($server) == 0) {
	$server = "127.0.0.1";
}

$unique = time();

$htdb_conn = pg_connect ("host=localhost dbname=htdb user=htdb password=htdb");

$kill_command = "/var/www/html/htdb/kill_live_tv.py";
$output = shell_exec($kill_command);

$query = "select hd.ipaddress from hdhomerun_channels as hc, hdhomerun_devices as hd where hc.channelid = $station and hc.deviceid = hd.deviceid and hd.tuner_used < hd.tunercount;";
$result = pg_query($htdb_conn, $query);
$rows = pg_num_rows($result);
for ($row = 0; $row < $rows; $row++ ){
	$record = pg_fetch_row($result, $row);
	$ip_addr = $record[0];
}

$query = "update hdhomerun_devices set tuner_used = tuner_used+1 where ipdaddress = '$ip_addr';";
//$result = pg_query($htdb_conn, $query);

$live_cmd = "/var/www/html/htdb/htdbtvlive.py $ip_addr tuner1 $station $unique  > /dev/null 2>/dev/null &";
//print "<!-- exec:$live_cmd -->\n";
$output = shell_exec($live_cmd);
//print "<!-- shell exec output = '$output' -->\n";
//exec($live_cmd);

$file_path = "/var/www/html/htdb/htdb-tvlive/hdhomerun_"."$station"."_"."$unique".".m3u8";
$url_path = "http://$server/htdb/htdb-tvlive/hdhomerun_"."$station"."_"."$unique".".m3u8";

$showtitle = "Channel $station";
$query = "select title, starttime, endtime from tv_info where channelid = $station and endtime > now() order by starttime limit 1;";
$result = pg_query($htdb_conn, $query);
$rows = pg_num_rows($result);
for ($row = 0; $row < $rows; $row++ ) {
	$record = pg_fetch_row($result, $row);
	$showtitle = $record[0];
}

$check = file_exists($file_path);
while ($check == 0) {
	sleep(1);
	$check = file_exists($file_path);
}

$query = "select nextval('tvliveid');";
$result = pg_query($htdb_conn, $query);
$record = pg_fetch_row($result, 0);
$id = $record[0];

$query = "insert into tv_live (station, filename, id) values ('$station', '$file_path', '$id');";
$result = pg_query($htdb_conn, $query);

pg_close($htdb_conn);

$old_file = "/var/www/html/htdb/htdbtvlive_template.template";
$myfile = fopen($old_file, "r");
$buffer = fread($myfile, filesize($old_file));
fclose($myfile);

$buffer = str_replace("<server>", $server, $buffer);
$buffer = str_replace("<mvfile>", $url_path, $buffer);
$buffer = str_replace("<showtitle>", $showtitle, $buffer);
$buffer = str_replace("<id>", $id, $buffer);

$new_file = "/var/www/html/htdb/htdb-tvlive/htdbtvlive_$station".".html";
$del_cmd = "/bin/rm -rf $new_file";
//print "<!-- exec:$del_cmd -->\n";
$output = shell_exec($del_cmd);

$myfile = fopen($new_file, "w");
fwrite($myfile,$buffer);
fclose($myfile);

if (strlen($cookie_client) > 0) {
	$query = "insert into client_playing (client, ttype, id) values ('$cookie_client', 'tvchannel', '$station');";
	print "<!-- query '$query' -->\n";
	$result = pg_query($htdb_conn, $query);
}

$header_url = "http://$server/htdb/htdb-tvlive/htdbtvlive_$station".".html";

//print "<!-- new url '$header_url' -->\n";
header("Location: $header_url");

?>
